Blog post on Parquet filter pushdown (#61)
apache/datafusion-site: content/blog/2025-03-21-parquet-pushdown.md

WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
{% endcomment %}
-->

_Editor's Note: This blog was first published on [Xiangpeng Hao's blog]. Thanks to [InfluxData] for sponsoring this work as part of his PhD funding._

[Xiangpeng Hao's blog]: https://blog.xiangpeng.systems/posts/parquet-pushdown/
[InfluxData]: https://www.influxdata.com/
<hr/>

In the [previous post], we discussed how [Apache DataFusion] prunes [Apache Parquet] files to skip irrelevant **files/row_groups** (sometimes also [pages](https://parquet.apache.org/docs/file-format/pageindex/)).

This post discusses how Parquet readers skip irrelevant **rows** while scanning data.

[previous post]: https://datafusion.apache.org/blog/2025/03/20/parquet-pruning
[Apache DataFusion]: https://datafusion.apache.org/
[Apache Parquet]: https://parquet.apache.org/

## Why filter pushdown in Parquet? 

Below is a query that reads sensor data with filters on `date_time` and `location`:

```sql
SELECT val, location 
FROM sensor_data 
WHERE date_time > '2025-03-12' AND location = 'office';
```

> <img src="/blog/images/parquet-pushdown/pushdown-vs-no-pushdown.jpg" alt="Parquet pruning skips irrelevant files/row_groups, while filter pushdown skips irrelevant rows. Without filter pushdown, all rows from location, val, and date_time columns are decoded before `location='office'` is evaluated. Filter pushdown is especially useful when the filter is selective, i.e., removes many rows." width="80%" class="img-responsive">

Reviewer: either that filter should be > '2025-03-11' or >= '2025-03-12'

> In our setup, sensor data is aggregated by date — each day has its own Parquet file.

Reviewer 2: imho it would be easier to reader if you attach an image how data partitioned

DataFusion prunes the unneeded Parquet files, i.e., `2025-03-10/11.parquet`.

> Once the files to read are located, the [*current default implementation*](https://github.com/apache/datafusion/issues/3463) reads all the projected columns (`sensor_id`, `val`, and `location`) into Arrow RecordBatches, then applies the filters over `location` to get the final set of rows.

Reviewer 3: ```suggestion
Once the files to read are located, the [*DataFusion's current default implementation*](https://github.com/apache/datafusion/issues/3463) reads all the projected columns (`sensor_id`, `val`, and `location`) into Arrow RecordBatches, then applies the filters over `location` to get the final set of rows.
```

> A better approach is **filter pushdown**, which evaluates filter conditions first and only decodes data that passes these conditions.

Reviewer 3: ```suggestion
A better approach is called **filter pushdown**, which evaluates filter conditions first and only decodes data that passes these conditions.
```

> In practice, this works by first processing only the filter columns (like `location`), building a boolean mask of rows that satisfy our conditions, then using this mask to selectively decode only the relevant rows from other columns (`sensor_id`, `val`). 

Reviewer 3: ```suggestion
In practice, this works by first processing only the filter columns (`date_time` and `location`), building a boolean mask of rows that satisfy our conditions, then using this mask to selectively decode only the relevant rows from other columns (`sensor_id`, `val`). 
```

> This eliminates the waste of decoding rows that will be filtered out.

Reviewer 3: ```suggestion
This eliminates the waste of decoding rows that will be immediately filtered out.
```

While simple in theory, practical implementations often make performance worse.

> ## Why slower?

Reviewer 3: ```suggestion
## How can filter pushdown be slower?
```

At a high level, the Parquet reader first builds a filter mask -- essentially a boolean array indicating which rows meet the filter criteria -- and then uses this mask to selectively decode only the needed rows from the remaining columns in the projection.

> Let's dig into details of [how filter pushdown is implemented](https://github.com/apache/arrow-rs/blob/d5339f31a60a4bd8a4256e7120fe32603249d88e/parquet/src/arrow/async_reader/mod.rs#L618-L712) in the current Rust implementation of Parquet readers.

Reviewer 3: ```suggestion
Let's dig into details of [how filter pushdown is implemented](https://github.com/apache/arrow-rs/blob/d5339f31a60a4bd8a4256e7120fe32603249d88e/parquet/src/arrow/async_reader/mod.rs#L618-L712) in the current Rust Parquet reader implementation, illustrated in the following figure.
```

The filter pushdown has two phases:

1. Build the filter mask (steps 1-3)

2. Apply the filter mask to the other columns (steps 4-7)

Within each phase, it takes three steps from Parquet to Arrow:

1. Decompress the Parquet pages using generic decompression algorithms like LZ4, Zstd, etc. (steps 1, 4, 6)

2. Decode the page content into Arrow format (steps 2, 5, 7)

3. Evaluate the filter over Arrow data (step 3)

In the figure above, we can see that `location` is **decompressed and decoded twice**, first when building the filter mask (steps 1, 2), and second when building the output (steps 4, 5).
This happens for all columns that appear both in the filter and output.

The table below shows the corresponding CPU time on the [ClickBench query 22](https://github.com/apache/datafusion/blob/main/benchmarks/queries/clickbench/queries.sql#L23):

```
+------------+--------+-------------+--------+
| Decompress | Decode | Apply filter| Others |
+------------+--------+-------------+--------+
| 206 ms     | 117 ms | 22 ms       | 48 ms  |
+------------+--------+-------------+--------+
```

> Clearly, decompress/decode operations dominate the time spent. With filter pushdown, we need to decompress/decode three times; but without filter pushdown, we only need to do this twice.

Reviewer 3: Why do we have to decompress/decode *three* times when pushdown is not on? It seems like the sentence should read

```suggestion
Clearly, decompress/decode operations dominate the time spent. With filter pushdown, we need to decompress/decode twice; but without filter pushdown, we only need to do this once.
```

> This explains why filter pushdown is slower.

Reviewer 3: ```suggestion
This explains why filter pushdown is slower in some cases.
```

> > **Note:** Highly selective filters may skip the entire page; but as long as we read one row from the page, we need to decompress/decode the entire page.

Reviewer 3: ```suggestion
> **Note:** Highly selective filters may skip the entire page; but as long as we read one row from the page, we need to decompress and often decode the entire page.
```

If you have to decode the entire page depends on the encoding you are using I think -- also I would love to start planting seeds that many of the supposed shortcomings of parquet the format are really short comings of the current encodings

## Attempt: cache filter columns

Intuitively, caching the filter columns and reusing them later could help.

> But caching consumes prohibitively high memory:

Reviewer 3: ```suggestion
But naively caching decoded pages consumes prohibitively high memory:
```

1. We need to cache Arrow arrays, which are on average [4x larger than Parquet data](https://github.com/XiangpengHao/liquid-cache/blob/main/dev/doc/liquid-cache-vldb.pdf).

> 2. We need to cache the **entire column in memory**, because in Phase 1 we build filters over the entire column, and only use it in Phase 2.  

Reviewer 3: I think technically it is the entire "column chunk" (aka all the values in the column for a particular row group)

3. The memory usage is proportional to the number of filter columns, which can be unboundedly high. 

> Worse, caching filter columns means we need to read partially from Parquet and partially from cache, which is complex to implement and requires a radical change to the current implementation. 

Reviewer 3: ```suggestion
Worse, caching filter columns means we need to read partially from Parquet and partially from cache, which is complex to implement, likely requiring a substantial change to the current implementation. 
```

> **Feel the complexity:** consider building a cache that properly handles nested columns, multiple filters, and filters with multiple columns.

## Real solution

We need a solution that:

1. Is simple to implement, i.e., doesn't require thousands of lines of code.

2. Incurs minimal memory overhead.

This section describes my [<700 LOC PR (with lots of comments and tests)](https://github.com/apache/arrow-rs/pull/6921#issuecomment-2718792433) that **reduces total ClickBench time by 15%, with up to 2x lower latency for some queries, no obvious regression on other queries, and caches at most 2 pages (~2MB) per column in memory**.

<img src="/blog/images/parquet-pushdown/new-pipeline.jpg" alt="New decoding pipeline, building filter mask and output columns are interleaved in a single pass, allowing us to cache minimal pages for minimal amount of time" width="80%" class="img-responsive">

The new pipeline interleaves the previous two phases into a single pass, so that:

1. The page being decompressed is immediately used to build filter masks and output columns.

> 2. We cache the decompressed page for minimal time; after one pass (steps 1-6), the cache memory is released for the next pass. 

Reviewer 2: We means Apache Datafusion?

This allows the cache to only hold 1 page at a time, and to immediately discard the previous page after it's used, significantly reducing the memory requirement for caching.

### What pages are cached?
You may have noticed that only `location` is cached, not `val`, because `val` is only used for output.
More generally, only columns that appear both in the filter and output are cached, and at most 1 page is cached for each such column.

More examples:
```sql
SELECT val 
FROM sensor_data 
WHERE date_time > '2025-03-12' AND location = 'office';
```

In this case, we don't cache any columns, because `val` is not used for filtering.

```sql
SELECT COUNT(*) 
FROM sensor_data 
WHERE date_time > '2025-03-12' AND location = 'office';
```

In this case, we also don't cache any columns, because the output projection is empty after query plan optimization.

> ### Then why cache 2 pages/column instead of 1? 

Reviewer 2: ```suggestion
### Then why cache 2 pages per column instead of 1? 
```

│ QQuery 35    │    700.79ms │     694.51ms │     no change │
│ QQuery 36    │    151.51ms │     152.93ms │     no change │
│ QQuery 37    │    108.18ms │      86.03ms │ +1.26x faster │
│ QQuery 38    │    114.64ms │     106.22ms │ +1.08x faster │
│ QQuery 39    │    260.80ms │     239.13ms │ +1.09x faster │
│ QQuery 40    │     60.74ms │      73.29ms │  1.21x slower │
│ QQuery 41    │     58.75ms │      67.85ms │  1.15x slower │
│ QQuery 42    │     65.49ms │      68.11ms │     no change │
└──────────────┴─────────────┴──────────────┴───────────────┘
┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━┓
┃ Benchmark Summary           ┃            ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━┩
│ Total Time (no-pushdown)    │ 38344.79ms │
│ Total Time (new-pushdown)   │ 32800.50ms │
│ Average Time (no-pushdown)  │   891.74ms │
│ Average Time (new-pushdown) │   762.80ms │
│ Queries Faster              │         13 │
│ Queries Slower              │          5 │
│ Queries with No Change      │         25 │
└─────────────────────────────┴────────────┘
```

## Conclusion

Despite being simple in theory, filter pushdown in Parquet is non-trivial to implement.

> It requires understanding both the Parquet format and reader implementation details. 
> The challenge lies in efficiently navigating through the dynamics of decoding, filter evaluation, and memory management.

Reviewer 3: Perhaps you can use this space to do some 🎣 (aka end with a "call to action")

```suggestion
It requires understanding both the Parquet format and reader implementation details. 
The challenge lies in efficiently navigating through the dynamics of decoding, filter evaluation, and memory management.

If you are interested in this level of optimization and want to help test, document and implement this type of optimization, come find us in the [DataFusion Community]. We would love to have you. 

[DataFusion Community]: https://datafusion.apache.org/contributor-guide/communication.html
```

│ QQuery 22    │   1927.94ms │    1404.35ms │ +1.37x faster │
│ QQuery 23    │   8104.11ms │    3610.76ms │ +2.24x faster │
│ QQuery 24    │    360.79ms │     330.55ms │ +1.09x faster │
│ QQuery 25    │    290.61ms │     252.54ms │ +1.15x faster │
│ QQuery 26    │    395.18ms │     362.72ms │ +1.09x faster │
│ QQuery 27    │    891.76ms │     959.39ms │  1.08x slower │
│ QQuery 28    │   4059.54ms │    4137.37ms │     no change │
│ QQuery 29    │    235.88ms │     228.99ms │     no change │
│ QQuery 30    │    564.22ms │     584.65ms │     no change │
│ QQuery 31    │    741.20ms │     757.87ms │     no change │
│ QQuery 32    │   2652.48ms │    2574.19ms │     no change │
│ QQuery 33    │   2373.71ms │    2327.10ms │     no change │
│ QQuery 34    │   2391.00ms │    2342.15ms │     no change │
│ QQuery 35    │    700.79ms │     694.51ms │     no change │
│ QQuery 36    │    151.51ms │     152.93ms │     no change │
│ QQuery 37    │    108.18ms │      86.03ms │ +1.26x faster │
│ QQuery 38    │    114.64ms │     106.22ms │ +1.08x faster │
│ QQuery 39    │    260.80ms │     239.13ms │ +1.09x faster │
│ QQuery 40    │     60.74ms │      73.29ms │  1.21x slower │

> │ QQuery 41    │     58.75ms │      67.85ms │  1.15x slower │

Reviewer 4: interesting that these are 15% and 21% slower, which is kind of significant. i guess these are <100ms while the queries that are sped up are taking 5x-100x longer

Reviewer 3: yeah or maybe the measurement error is high 

I think chasing down this kind of stuff is what we need to do to productionize the work

See the License for the specific language governing permissions and
limitations under the License.
{% endcomment %}
-->

_Editor's Note: This blog was first published on [Xiangpeng Hao's blog]. Thanks to [InfluxData] for sponsoring this work as part of his PhD funding._

[Xiangpeng Hao's blog]: https://blog.xiangpeng.systems/posts/parquet-pushdown/
[InfluxData]: https://www.influxdata.com/
<hr/>

In the [previous post], we discussed how [Apache DataFusion] prunes [Apache Parquet] files to skip irrelevant **files/row_groups** (sometimes also [pages](https://parquet.apache.org/docs/file-format/pageindex/)).

This post discusses how Parquet readers skip irrelevant **rows** while scanning data,
leveraging Parquet's columnar layout by first reading only filter columns,
and then selectively reading other columns only for matching rows.

[previous post]: https://datafusion.apache.org/blog/2025/03/20/parquet-pruning
[Apache DataFusion]: https://datafusion.apache.org/
[Apache Parquet]: https://parquet.apache.org/

## Why filter pushdown in Parquet? 

Below is an example query that reads sensor data with filters on `date_time` and `location`.

> Without filter pushdown, all rows from location, val, and date_time columns are decoded before `location='office'` is evaluated. Filter pushdown is especially useful when the filter is selective, i.e., removes many rows.

Reviewer 5: ```suggestion
Without filter pushdown, all rows from `location`, `val`, and `date_time` columns are decoded before `location='office'` is evaluated. Filter pushdown is especially useful when the filter is selective, i.e., removes many rows.
```

  <figcaption>
    Parquet pruning skips irrelevant files/row_groups, while filter pushdown skips irrelevant rows. Without filter pushdown, all rows from location, val, and date_time columns are decoded before `location='office'` is evaluated. Filter pushdown is especially useful when the filter is selective, i.e., removes many rows.
  </figcaption>
</figure>

In our setup, sensor data is aggregated by date — each day has its own Parquet file.
At planning time, DataFusion prunes the unneeded Parquet files, i.e., `2025-03-10.parquet` and `2025-03-11.parquet`.

Once the files to read are located, the [*DataFusion's current default implementation*](https://github.com/apache/datafusion/issues/3463) reads all the projected columns (`sensor_id`, `val`, and `location`) into Arrow RecordBatches, then applies the filters over `location` to get the final set of rows.

> A better approach is called **filter pushdown**, which evaluates filter conditions first and only decodes data that passes these conditions.

Reviewer 5: ```suggestion
A better approach is called **filter pushdown** with **late materialization**, which evaluates filter conditions first and only decodes data that passes these conditions.
```

  <figcaption>
    New decoding pipeline, building filter mask and output columns are interleaved in a single pass, allowing us to cache minimal pages for minimal amount of time
  </figcaption>
</figure>

The new pipeline interleaves the previous two phases into a single pass, so that:

1. The page being decompressed is immediately used to build filter masks and output columns.

2. Decompressed pages are cached for minimal time; after one pass (steps 1-6), the cache memory is released for the next pass. 

This allows the cache to only hold 1 page at a time, and to immediately discard the previous page after it's used, significantly reducing the memory requirement for caching.

### What pages are cached?
You may have noticed that only `location` is cached, not `val`, because `val` is only used for output.
More generally, only columns that appear both in the filter and output are cached, and at most 1 page is cached for each such column.

More examples:
```sql
SELECT val 
FROM sensor_data 

> WHERE date_time > '2025-03-12' AND location = 'office';

Reviewer 4: nit: match above
```suggestion
WHERE date_time > '2025-03-11' AND location = 'office';
```

```

In this case, no columns are cached, because `val` is not used for filtering.

```sql
SELECT COUNT(*) 
FROM sensor_data 

> WHERE date_time > '2025-03-12' AND location = 'office';

Reviewer 4: nit: match above
```suggestion
WHERE date_time > '2025-03-11' AND location = 'office';
```

  </figcaption>
</figure>

This means that to decode a page of data, it actually references two pages: the dictionary page and the data page.

This is why it caches 2 pages per column: one dictionary page and one data page.
The data page slot will move forward as it reads the data; but the dictionary page slot always references the first page.

<figure>
  <img src="/blog/images/parquet-pushdown/cached-pages.jpg" alt="Cached two pages, one for dictionary (pinned), one for data (moves as it reads the data)" width="80%" class="img-responsive">
  <figcaption>
    Cached two pages, one for dictionary (pinned), one for data (moves as it reads the data)
  </figcaption>
</figure>

## How does it perform?

Here are my results on [ClickBench](https://github.com/apache/datafusion/tree/main/benchmarks#clickbench) on my AMD 9900X machine. The total time is reduced by 15%, with Q23 being 2.24x faster,
and queries that get slower are likely due to noise.

```
┏━━━━━━━━━━━━━━┳━━━━━━━━━━━━━┳━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Query        ┃ no-pushdown ┃ new-pushdown ┃        Change ┃
┡━━━━━━━━━━━━━━╇━━━━━━━━━━━━━╇━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩

> │ QQuery 0     │      0.47ms │       0.43ms │ +1.10x faster │

Reviewer 6: Any thoughts on using percent to communicate difference in performance. 
`+1.10x faster` is not immediately obvious is the same as `10% faster`.

Author: I used this script to generate the table: https://github.com/apache/datafusion/blob/main/benchmarks/compare.py

I agree the exact presentation can be improved, but it is consistent with other datafusion benchmark reports.
